File: solutions/csharp/zebra-puzzle/3/ZebraPuzzle.cs
public enum Color { Red, Green, Ivory, Yellow, Blue }
public enum Nationality { Englishman, Spaniard, Ukrainian, Japanese, Norwegian }
public enum Pet { Dog, Snails, Fox, Horse, Zebra }
public enum Drink { Coffee, Tea, Milk, OrangeJuice, Water }
public enum Smoke { OldGold, Kools, Chesterfields, LuckyStrike, Parliaments }

public static class ZebraPuzzle
{
    public static Nationality DrinksWater() => NationalityMatching(house => house.Drink == Drink.Water);

    public static Nationality OwnsZebra() => NationalityMatching(house => house.Pet == Pet.Zebra);

    private static Nationality NationalityMatching(Func<House, bool> predicate) =>
        FindHousesMatchingAllRules()
            .First(predicate)
            .Nationality;

    private static List<House> FindHousesMatchingAllRules() =>
        (
            from colors in ValidColors()
            from nationalities in MatchingNationalities(colors)
            from drinks in MatchingDrinks(nationalities)
            from smokes in MatchingSmokes(drinks)
            from pets in MatchingPets(smokes)
            select pets
        ).First();

    private static IEnumerable<ColorSolution> ValidColors() =>
        Permutations<Color>()
            .Satisfying(GreenImmediatelyRightOfIvory)
            .Satisfying(BlueIsSecond)
            .Select(colors => new ColorSolution(colors));

    private static IEnumerable<NationalitySolution> MatchingNationalities(ColorSolution solution) =>
        Permutations<Nationality>()
            .Satisfying(NorwegianLivesInFirstHouse)
            .Satisfying(EnglishmanLivesInRedHouse(solution))
            .Select(nationalities => new NationalitySolution(solution.Colors, nationalities));

    private static IEnumerable<DrinkSolution> MatchingDrinks(NationalitySolution solution) =>
        Permutations<Drink>()
            .Satisfying(MilkInMiddleHouse)
            .Satisfying(GreenHouseDrinksCoffee(solution))
            .Satisfying(UkrainianDrinksTea(solution))
            .Select(drinks => new DrinkSolution(solution.Colors, solution.Nationalities, drinks));

    private static IEnumerable<SmokeSolution> MatchingSmokes(DrinkSolution solution) =>
        Permutations<Smoke>()
            .Satisfying(PainterLivesInYellowHouse(solution))
            .Satisfying(FootballPlayerDrinksOrangeJuice(solution))
            .Satisfying(JapanesePersonPlaysChess(solution))
            .Select(smokes => new SmokeSolution(solution.Colors, solution.Nationalities, solution.Drinks, smokes));

    private static IEnumerable<List<House>> MatchingPets(SmokeSolution solution) =>
        Permutations<Pet>()
            .Satisfying(SpaniardOwnsDog(solution))
            .Satisfying(SnailOwnerGoesDancing(solution))
            .Satisfying(ReaderLivesNextToFox(solution))
            .Satisfying(PainterLivesNextToHorse(solution))
            .Select(pets => BuildHouses(solution.Colors, solution.Nationalities, pets, solution.Drinks, solution.Smokes));

    private static IEnumerable<T[]> Satisfying<T>(this IEnumerable<T[]> candidates, Func<T[], bool> rule) =>
        candidates.Where(rule);

    private static bool GreenImmediatelyRightOfIvory(Color[] colors) =>
        Position(colors, Color.Green) == Position(colors, Color.Ivory) + 1;

    private static bool BlueIsSecond(Color[] colors) => Position(colors, Color.Blue) == 1;

    private static bool NorwegianLivesInFirstHouse(Nationality[] nationalities) =>
        Position(nationalities, Nationality.Norwegian) == 0;

    private static Func<Nationality[], bool> EnglishmanLivesInRedHouse(ColorSolution solution) =>
        nationalities => SamePosition(nationalities, Nationality.Englishman, solution.Colors, Color.Red);

    private static bool MilkInMiddleHouse(Drink[] drinks) => Position(drinks, Drink.Milk) == 2;

    private static Func<Drink[], bool> GreenHouseDrinksCoffee(NationalitySolution solution) =>
        drinks => SamePosition(drinks, Drink.Coffee, solution.Colors, Color.Green);

    private static Func<Drink[], bool> UkrainianDrinksTea(NationalitySolution solution) =>
        drinks => SamePosition(drinks, Drink.Tea, solution.Nationalities, Nationality.Ukrainian);

    private static Func<Smoke[], bool> PainterLivesInYellowHouse(DrinkSolution solution) =>
        smokes => SamePosition(smokes, Smoke.Kools, solution.Colors, Color.Yellow);

    private static Func<Smoke[], bool> FootballPlayerDrinksOrangeJuice(DrinkSolution solution) =>
        smokes => SamePosition(smokes, Smoke.LuckyStrike, solution.Drinks, Drink.OrangeJuice);

    private static Func<Smoke[], bool> JapanesePersonPlaysChess(DrinkSolution solution) =>
        smokes => SamePosition(smokes, Smoke.Parliaments, solution.Nationalities, Nationality.Japanese);

    private static Func<Pet[], bool> SpaniardOwnsDog(SmokeSolution solution) =>
        pets => SamePosition(pets, Pet.Dog, solution.Nationalities, Nationality.Spaniard);

    private static Func<Pet[], bool> SnailOwnerGoesDancing(SmokeSolution solution) =>
        pets => SamePosition(pets, Pet.Snails, solution.Smokes, Smoke.OldGold);

    private static Func<Pet[], bool> ReaderLivesNextToFox(SmokeSolution solution) =>
        pets => NextTo(pets, Pet.Fox, solution.Smokes, Smoke.Chesterfields);

    private static Func<Pet[], bool> PainterLivesNextToHorse(SmokeSolution solution) =>
        pets => NextTo(pets, Pet.Horse, solution.Smokes, Smoke.Kools);

    private static List<House> BuildHouses(Color[] colors, Nationality[] nationalities, Pet[] pets, Drink[] drinks, Smoke[] smokes) =>
        Enumerable
          .Range(0, 5)
          .Select(position => new House(colors[position], nationalities[position], pets[position], drinks[position], smokes[position]))
          .ToList();

    private static bool NextTo(int left, int right) => Math.Abs(left - right) == 1;

    private static bool NextTo<TLeft, TRight>(TLeft[] leftValues, TLeft left, TRight[] rightValues, TRight right) =>
        NextTo(Position(leftValues, left), Position(rightValues, right));

    private static bool SamePosition<TLeft, TRight>(TLeft[] leftValues, TLeft left, TRight[] rightValues, TRight right) =>
        Position(leftValues, left) == Position(rightValues, right);

    private static int Position<T>(T[] values, T value) => Array.IndexOf(values, value);

    private static IEnumerable<T[]> Permutations<T>() where T : struct, Enum => Permutations(Enum.GetValues<T>());

    private static IEnumerable<T[]> Permutations<T>(T[] values) =>
        values.Length == 1
            ? [values]
            : values.SelectMany(value =>
                Permutations(values.Where(other => !other!.Equals(value)).ToArray()),
                (value, permutation) => new[] { value }.Concat(permutation).ToArray());

    private record ColorSolution(Color[] Colors);

    private record NationalitySolution(Color[] Colors, Nationality[] Nationalities);

    private record DrinkSolution(Color[] Colors, Nationality[] Nationalities, Drink[] Drinks);

    private record SmokeSolution(Color[] Colors, Nationality[] Nationalities, Drink[] Drinks, Smoke[] Smokes);

    private record House(Color Color, Nationality Nationality, Pet Pet, Drink Drink, Smoke Smoke);

}

File: solutions/scala/zebra-puzzle/4/src/main/scala/ZebraPuzzle.scala
object ZebraPuzzle {

  sealed trait Resident
  case object Englishman extends Resident
  case object Spaniard extends Resident
  case object Ukrainian extends Resident
  case object Norwegian extends Resident
  case object Japanese extends Resident

  case class Solution(waterDrinker: Resident, zebraOwner: Resident)

  lazy val solve: Solution =
    solutions.next()

  private enum Nation:
    case Englishman, Spaniard, Ukrainian, Norwegian, Japanese

  private enum Color:
    case Red, Green, Ivory, Yellow, Blue

  private enum Smoke:
    case OldGold, Kools, Chesterfields, LuckyStrike, Parliaments

  private enum Pet:
    case Dog, Snails, Fox, Horse, Zebra

  private enum Drink:
    case Coffee, Tea, Milk, OrangeJuice, Water

  private def solutions: Iterator[Solution] =
    for
      nations <- permutations(Nation.values)
      if nations.indexOf(Nation.Norwegian) == 0

      colors <- permutations(Color.values)
      if nations.indexOf(Nation.Englishman) == colors.indexOf(Color.Red)
      if nextTo(nations.indexOf(Nation.Norwegian), colors.indexOf(Color.Blue))
      if colors.indexOf(Color.Ivory) + 1 == colors.indexOf(Color.Green)

      drinks <- permutations(Drink.values)
      if drinks.indexOf(Drink.Milk) == 2
      if nations.indexOf(Nation.Ukrainian) == drinks.indexOf(Drink.Tea)
      if colors.indexOf(Color.Green) == drinks.indexOf(Drink.Coffee)

      pets <- permutations(Pet.values)
      if nations.indexOf(Nation.Spaniard) == pets.indexOf(Pet.Dog)

      smokes <- permutations(Smoke.values)
      if smokes.indexOf(Smoke.OldGold) == pets.indexOf(Pet.Snails)
      if smokes.indexOf(Smoke.Kools) == colors.indexOf(Color.Yellow)
      if smokes.indexOf(Smoke.LuckyStrike) == drinks.indexOf(Drink.OrangeJuice)
      if smokes.indexOf(Smoke.Parliaments) == nations.indexOf(Nation.Japanese)
      if nextTo(smokes.indexOf(Smoke.Chesterfields), pets.indexOf(Pet.Fox))
      if nextTo(smokes.indexOf(Smoke.Kools), pets.indexOf(Pet.Horse))
    yield Solution(
      waterDrinker = residentFor(nations(drinks.indexOf(Drink.Water))),
      zebraOwner = residentFor(nations(pets.indexOf(Pet.Zebra)))
    )

  private def permutations[A](values: Array[A]): Iterator[List[A]] =
    values.toList.permutations

  private def nextTo(left: Int, right: Int): Boolean =
    (left - right).abs == 1

  private def residentFor(nation: Nation): Resident =
    nation match
      case Nation.Englishman => Englishman
      case Nation.Spaniard => Spaniard
      case Nation.Ukrainian => Ukrainian
      case Nation.Norwegian => Norwegian
      case Nation.Japanese => Japanese
}
